Compose: Add DaxTextField

landomen opened the pull request:
Task/Issue URL: https://app.asana.com/1/137249556945/project/1210594645151737/task/1211659112661196?focus=true

### Description
Adds a Compose version of the DaxTextField.

### Steps to test this PR

_ADS Preview_
- [ ] Go to ADS Preview screen and click on Text Input tab 
- [ ] Check compose versions 

### UI changes
See ADS Preview screen

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces Compose text fields (standard and secure) with trailing icon API and previews, adds theme text-field colors, and enforces trailing icon usage via new lint checks.
> 
> - **Design System (Compose)**:
>   - **New components**: `DaxTextField` and `DaxSecureTextField` with `DaxTextFieldInputMode`, `DaxTextFieldLineLimits`, keyboard options, input transformations, error/supporting text, and `DaxTextFieldTrailingIconScope` API.
>   - **Secure field**: visibility toggle integrated with optional trailing icon.
>   - **Previews**: comprehensive previews for states and configurations.
> - **ADS Preview/Examples**:
>   - Integrates multiple Compose samples in `ComponentTextInputFragment` and `component_text_input_view.xml` (hint, single/multi/form, read-only, disabled, error, URL/IP, focus behaviors, observable text, password variants).
> - **Theme**:
>   - Adds `DuckDuckGoTextFieldColors` (`colors.textField.borders`) and exposes `DuckDuckGoTheme.iconColors`; minor color addition `White78`.
> - **Lint Rules**:
>   - New detectors: `DaxTextFieldTrailingIconDetector` and `DaxSecureTextFieldTrailingIconDetector`; registered in `DuckDuckGoIssueRegistry` with extensive tests to enforce using `DaxTextFieldTrailingIconScope` in `trailingIcon`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 33aa9f12d2ecac08e12727d003e5366c74f59633. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

landomen commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxSecureTextField.kt:
```diff
@@ -0,0 +1,388 @@
+ * Figma reference: https://www.figma.com/design/BOHDESHODUXK7wSRNBOHdu/%F0%9F%A4%96-Android-Components?m=auto&node-id=3202-5150
+ */
+@Composable
+fun DaxSecureTextField(
```
This needs to be a separate composable as it's using a `OutlinedSecureTextField` underneath instead of the regular `OutlinedTextField`

malmstein commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,562 @@
+        MultiLine -> TextFieldLineLimits.MultiLine(minHeightInLines = 1)
+        Form -> TextFieldLineLimits.MultiLine(minHeightInLines = 3)
+    }
+}
```
I prefer having one instead, good call!

mikescamell replied:
Yeah I agree with the enum approach in this case specifically, it's the same InputText with just different line limits. Usually I'd move towards what Gerard suggested but in this case I think it makes more sense. We'll see more of this when I raise the DaxButton PR 👀

GerardPaligot commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,562 @@
+fun DaxTextField(
+    state: TextFieldState,
+    modifier: Modifier = Modifier,
+    hint: String? = null,
```
Warning: We need to take care about the naming here. The Compose TextField has two native parameters: `placeholder` and `label`. `placeholder` is displayed when you click on the text field but the user didn't enter anything yet and `label` is displayed at the beginning in the text field and is pushed at the end when the user click on the component.

`hint` seems too vague and according to the Figma specification, it seems that the `placeholder` isn't documented in our component specification. Maybe we should ask to a designer what they want and impact the api contract here according to their feedback?

mikescamell replied:
Yeah tough one seeing as the Design System has nothing in this regard except "label". I'm open to having both label and placeholder as we moving away from XML, but I understand why Domen went for this due how it's used. I'm not too concerned about forgetting XML property names, Domen has documented how hint behaves in the kdoc and in my opinion we should build Composables to match our design system components where possible, and not Google's Material APIs. I lean towards label and placeholder I think in this case 👍

landomen replied:
I agree with both of you. Renamed the parameter from `hint` to `label` to follow the standard Compose naming. 

I'm not exposing a `placeholder` parameter at this time, as it conflicts with how our existing ADS text input currently behaves (see video, the placeholder is constantly visible). It should be pretty simple to expose it later if we need it.

Let me know if you still think it's a good idea to expose it, I can add it.

https://github.com/user-attachments/assets/d1a3293c-2c33-4be8-8689-6e575bec3607

GerardPaligot commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/theme/Theme.kt:
```diff
@@ -81,6 +81,7 @@ fun DuckDuckGoTheme(
         window = colorResource(R.color.white),
         destructive = colorResource(R.color.alertRedOnLightDefault),
         lines = colorResource(R.color.black9),
+        borders = colorResource(R.color.black30),
```
It depends a lot on the design spec. This `borders` is used only in the text field or every time we need to use a border, they'll use the exact same border color?

GerardPaligot commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,562 @@
+    trailingIcon: DaxTextFieldTrailingIcon? = null,
+    keyboardOptions: KeyboardOptions = DaxTextFieldDefaults.TextKeyboardOptions,
+    inputTransformation: InputTransformation? = null,
+    onClick: (() -> Unit)? = null,
```
The need behind the click on the text field seems to be more linked to the focus than something else, so the need behind this click can be covered by this https://github.com/duckduckgo/Android/pull/7247#discussion_r2603559243? What do you think?

About api slot or not and with a scope or not, it is always hard to take this kind of decision in compose. For example, we can suggest this kind of contract:

```
@Composable
fun DaxTextField(
    state: TextFieldState,
    modifier: Modifier = Modifier,
    hint: String? = null,
    lineLimits: DaxTextFieldLineLimits = DaxTextFieldLineLimits.MultiLine,
    enabled: Boolean = true,
    editable: Boolean = true,
    clickable: Boolean = false,
    error: String? = null,
    trailingIcon: (@Composable TrailingScope.() -> Unit)? = null,
    keyboardOptions: KeyboardOptions = DaxTextFieldDefaults.TextKeyboardOptions,
    inputTransformation: InputTransformation? = null,
    interactionSource: MutableInteractionSource = remember { MutableInteractionSource() },
    // onClick: (() -> Unit)? = null, With interactionSource and api slot trailingIcon, we don't need this parameter anymore
)

object TrailingScope {
    @Composable
    fun IconButton(
        @DrawableRes iconResId: Int,
        contentDescription: String?,
        modifier: Modifier = Modifier,
        tint: Color = DuckDuckGoTheme.colors.iconPrimary,
        enabled: Boolean = true,
        onClick: (() -> Unit)? = null,
    ) {
      // expected composable here
    }
}
```

Then, you can use the component like this:

```
val state = rememberTextFieldState("This is not editable.")
DaxTextField(
    state = state,
    hint = "Non-editable text with end icon",
    trailingIcon = {
        IconButton(
            iconResId = R.drawable.ic_copy_24,
            contentDescription = "Copy",
            onClick = { toastOnClick(Action.PerformEndAction) }
        )
    },
    editable = false,
    lineLimits = DaxTextFieldLineLimits.SingleLine,
)
```

This is useful if you want to help the component consumer to declare the correct component in the api slot but it isn't required. If for some reasons, you want something else, scoped api slot doesn't force you to use one component present in the scope object.

This is also possible:

```
DaxTextField(
    state = state,
    hint = "Non-editable text with end icon",
    trailingIcon = {
        Text(text = "My end icon")
    },
    editable = false,
    lineLimits = DaxTextFieldLineLimits.SingleLine,
)
```

So yes, if we chose the api slot contract and we won't this kind of thing in the codebase, we'll need to write linter rules. (btw, we'll probably need to write linter rules when we'll start to use compose in the codebase for many other usecases).

landomen replied:
Thanks for the suggestion, @GerardPaligot! I went with the scoped api slot. I'll write a lint rule to ensure we're only using the expected composable. We already have a lint rule for DaxText that was written by @mikescamell .

GerardPaligot commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,562 @@
+    lineLimits: DaxTextFieldLineLimits = DaxTextFieldLineLimits.MultiLine,
+    enabled: Boolean = true,
+    editable: Boolean = true,
+    clickable: Boolean = false,
```
Let me build the state matrix here to be sure I'm understanding everything well:

Usecases | Enabled | Editable | Clickable
----- | -------- | -------- | ----------
I can fully interact with the text field | x | x | x
Read only mode | x | | x
Edited by the business logic, not by user |  | x |  x
**No use case for this scenario** | x | x | 
**No use case for this scenario** | x |  | 
Edited by the business logic, not by user | | x | 
Focusable mode | | | x

Am I right?

landomen replied:
Decided to go with an enum for simplicity with 3 values: `Editable`, `ReadOnly` and `Disabled`. I initially also had `Clickable` but it resulted in the same combination as `ReadOnly` (`enabled=true`, `readOnly =true`). Since we'll be using an InteractionSource now to detect focus and clicks, it's not really needed. Consumers can simply set the text field to ReadOnly and apply the interaction source.

mikescamell commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/theme/Color.kt:
```diff
@@ -40,11 +40,13 @@ data class DuckDuckGoColors(
     val accentYellow: Color,
     val ripple: Color,
     val text: DuckDuckGoTextColors,
+    val iconPrimary: Color,
```
💡Could we already create a DuckDuckGoIconColors with iconPrimary as our first? Similar to DuckDuckGoTextColors

landomen replied:
Good idea! Introduced `DuckDuckGoIconColors`

GerardPaligot commented on android-design-system/design-system-internal/src/main/java/com/duckduckgo/common/ui/internal/ui/component/textinput/ComponentTextInputFragment.kt:
```diff
@@ -58,11 +96,429 @@ class ComponentTextInputFragment : Fragment() {
+        }
+
+        // Error
+        view.setupThemedComposeView(id = com.duckduckgo.common.ui.internal.R.id.compose_text_input_21, isDarkTheme = isDarkTheme) {
```
We have a difference between the XML and Compose components here.
When I'm clicking on the XML view, the cursor is white/black (according to if you are in dark/light mode).
When I'm clicking on the Compose view, the cursor is in red.
From my opinion, we should keep the Compose behavior but because there is a difference, I want to report it here.

landomen commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxSecureTextField.kt:
```diff
@@ -0,0 +1,385 @@
+) {
+    // needed by the OutlinedTextField container
+    val internalInteractionSource = interactionSource ?: remember { MutableInteractionSource() }
+    var isPasswordVisible by remember { mutableStateOf(false) }
```
Lifted the state outside so it's now the consumers responsibility. It makes sense to have a stateless component, however I do think this impacts the ease of use by needing more boilerplate state management. 

In all the currently supported use-cases , the only reason to show/hide password is because of the internal icon click. It seems we're predicting potential other usages in the future, incorporating that in the API, and by doing that,  making it harder to use for existing purpose.

mikescamell replied:
Could we consider creating a stateful version of this Composable alongside this one? That way we have one where callers can deal with the state and one where callers don't care and it just shows/hides.

We could even make DaxSecureTextField internal for the time being, and then if there is a need to expose this version in future when we eventually migrate then we can consider it.

landomen commented on android-design-system/design-system-internal/src/main/java/com/duckduckgo/common/ui/internal/ui/component/textinput/ComponentTextInputFragment.kt:
```diff
@@ -58,11 +96,429 @@ class ComponentTextInputFragment : Fragment() {
+        }
+
+        // Non-editable text full click listener with end icon
+        view.setupThemedComposeView(id = com.duckduckgo.common.ui.internal.R.id.compose_text_input_30, isDarkTheme = isDarkTheme) {
```
Not sure if there is an easy way to prevent that. Would probably have to do something with the `MutableInteractionSource` but that could be tricky. In my opinion, the current behavior is okay, because you interacted with the text field (it's still enabled). 

The XML one sometimes also behaves like that, at least on the "Non-editable text with line truncation" example. And there is a bug on the XML one where the keyboard still shows up and you can enter new text even though its set to non-editable 😅

mikescamell replied:
SGTM @landomen, lets go with it 👍

landomen commented:
@GerardPaligot @mikescamell I pushed two new commits:
* [Lift password state - addresses comments](https://github.com/duckduckgo/Android/pull/7247/commits/9c4d4a6bb480ea0e5853fb0b0f47aa245a0120ad)
* [Add new lint detector - ensures we can only use `DaxTextFieldTrailingIcon` composable in the `trailingIcon` parameter](https://github.com/duckduckgo/Android/pull/7247/commits/464f7a685fb608bed905afdc59a01e41beda76ba)

mikescamell commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,540 @@
+}
+
+@Stable
+enum class DaxTextFieldInputMode {
```
❤️

GerardPaligot commented on lint-rules/src/main/java/com/duckduckgo/lint/ui/DaxSecureTextFieldTrailingIconDetector.kt:
```diff
@@ -0,0 +1,113 @@
+
+            // Only DaxTextFieldTrailingIcon should be used in the trailingIcon parameter
+            // If the source doesn't contain it, then it's invalid
+            return !source.contains("DaxTextFieldTrailingIcon")
```
We can probably get member list declared in the scope object class and check if the source contains one member of this list? This will allow us to avoid to edit this linter rule every time we need to edit the scope.

GerardPaligot commented on lint-rules/src/main/java/com/duckduckgo/lint/ui/DaxTextFieldTrailingIconDetector.kt:
```diff
@@ -0,0 +1,111 @@
+
+            // Only DaxTextFieldTrailingIcon should be used in the trailingIcon parameter
+            // If the source doesn't contain it, then it's invalid
+            return !source.contains("DaxTextFieldTrailingIcon")
```
We can probably get member list declared in the scope object class and check if the source contains one member of this list? This will allow us to avoid to edit this linter rule every time we need to edit the scope.

landomen commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,540 @@
+            ),
+        ),
+    ) {
+        CompositionLocalProvider(LocalTextSelectionColors provides daxTextFieldColors().textSelectionColors) {
```
Yes, good catch. Updated!

landomen commented on android-design-system/design-system-internal/src/main/java/com/duckduckgo/common/ui/internal/ui/component/textinput/ComponentTextInputFragment.kt:
```diff
@@ -58,11 +96,429 @@ class ComponentTextInputFragment : Fragment() {
+        // Editable password that fits in one line
+        view.setupThemedComposeView(id = com.duckduckgo.common.ui.internal.R.id.compose_text_input_6, isDarkTheme = isDarkTheme) {
+            val state = rememberTextFieldState("Loremipsumolor")
+            DaxSecureTextField(
```
@malmstein Any idea why the xml input field reveals the password automatically when clicked?

landomen commented on android-design-system/design-system/src/main/java/com/duckduckgo/common/ui/compose/textfield/DaxTextField.kt:
```diff
@@ -0,0 +1,546 @@
+
+    @Composable
+    fun SomeComposable() {
+    }
```
Removed

landomen commented:
@mikescamell @GerardPaligot I've addressed the open comments, please take another look

GerardPaligot reviewed (approved):
LGTM! 🎉 Congrats @landomen for this contribution
